Implement key-value pair iteration of dictionaries

rodrigodias4 opened the pull request:
Implementation includes parsing of both variables, type hints, basic static analysis and type checking, compiling, and runtime type-checking. The implementation writes an implicit assignment from the index (first for variable) of the container to the second for variable before the suite, within the compilation and code generation stages.
Added a new OPCODE (OPCODE_FOR_SECOND_VARIABLE_DICT_TYPE_TEST), used when a second variable is declared, to check if type of container is a dictionary at runtime before entering the loop.

Example grammar:
forStmt = "for" IDENTIFIER [ “:” typeHint] [ “,” IDENTIFIER [ “:” typeHint ]] "in" expression ":" stmtOrSuite ;

Unit tests added (.gd and .out):
* parser/features/for_key_value_pair : basic success test
* parser/features/for_key_value_pair_with_type_specifiers : successful key-value pair iteration on dictionary with type specifiers for both key and value variables
* analyzer/errors/for_key_value_not_dictionary : error on key-value pair iteration on non-dictionary container (static analysis at compile-time if container is statically typed)
* runtime/features/for_key_value_multiple_types : successful key-value pair iteration on dictionary with multiple types
* runtime/errors/for_key_value_not_dictionary : error on key-value pair iteration on non-dictionary container (runtime check on variant container)
* runtime/errors/for_key_value_incompatible_types : error on incompatible types between the static type hint and the container’s elements’ type

_Bugsquad edit:_ Closes godotengine/godot-proposals#3457.

rodrigodias4 committed:
Implement key-value pair iteration of dictionaries
Implementation includes parsing of both variables, type hints, basic static analysis and type checking, compiling, and runtime type-checking. The implementation writes an implicit assignment from the index (first for variable) of the container to the second for variable before the suite, within the compilation and code generation stages.
Added a new OPCODE (OPCODE_FOR_SECOND_VARIABLE_DICT_TYPE_TEST), used when a second variable is declared, to check if type of container is a dictionary at runtime before entering the loop.

Example grammar:
forStmt = "for" IDENTIFIER [ “:” typeHint] [ “,” IDENTIFIER [ “:” typeHint ]]” in" expression ":" stmtOrSuite ;

Unit tests added (.gd and .out):
* parser/features/for_key_value_pair : basic success test
* parser/features/for_key_value_pair_with_type_specifiers : successful key-value pair iteration on dictionary with type specifiers for both key and value variables
* analyzer/errors/for_key_value_not_dictionary : error on key-value pair iteration on non-dictionary container (static analysis at compile-time if container is statically typed)
* runtime/features/for_key_value_multiple_types : successful key-value pair iteration on dictionary with multiple types
* runtime/errors/for_key_value_not_dictionary : error on key-value pair iteration on non-dictionary container (runtime check on variant container)
* runtime/errors/for_key_value_incompatible_types : error on incompatible types between the static type hint and the container’s elements’ type

dalexeev commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -2122,6 +2122,7 @@ void GDScriptAnalyzer::resolve_for(GDScriptParser::ForNode *p_for) {
 	}
 
 	GDScriptParser::DataType variable_type;
+	GDScriptParser::DataType pair_variable_type;
```
```suggestion
	GDScriptParser::DataType first_variable_type;
	GDScriptParser::DataType second_variable_type;
```

`pair` is confusing because it is the type of the second variable, not the type of the pair as a whole. `key` and `value` won't work because arrays iterate over values. The same in other places.

dalexeev commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -2174,6 +2175,14 @@ void GDScriptAnalyzer::resolve_for(GDScriptParser::ForNode *p_for) {
+		if (p_for->pair_variable) {
+			if (list_type.builtin_type == Variant::DICTIONARY) {
+				pair_variable_type.kind = GDScriptParser::DataType::VARIANT;
+				mark_node_unsafe(p_for->pair_variable);
```
```suggestion
```

dalexeev reviewed (commented):
1\. In my opinion, the syntax should be `for key: KeyType, value: ValueType in iterable`. Because a. the type is not separated from the identifier, and b. `key: Type, value` is less confusing than `key, value: Type` if you omit the value's type hint.

2\. GDScript is a gradually typed programming language, so we do not require static typing anywhere.

![](https://github.com/godotengine/godot/assets/47700418/b63fb746-70cc-41a9-9d3e-02a8dfc6bf1c)

In case the iterable value is a supertype of `Dictionary` (i.e. `Variant`), there should be no compile-time error, only the second iterator variable should be marked as unsafe.

Type checking should be performed at runtime, if necessary. If the value turns out to be a dictionary, then ok. If not, then there must be a runtime error. You can implement this either in the compiler (by generating type checking code if necessary, much like we did for `match`) or by adding new opcode(s) for the VM.

3\. The type of the second iterator variable is not converted for implicitly convertible types (like `int` and `float`) and does not cause a runtime error for incompatible types.

![](https://github.com/godotengine/godot/assets/47700418/8e69b057-6686-4b15-a2d2-af523e85ee88)
![](https://github.com/godotengine/godot/assets/47700418/9b7f1ea3-4f76-48d3-a9ea-bc0c52e4c6e2)

This is inconsistent with the behavior of the first iterator variable.

![](https://github.com/godotengine/godot/assets/47700418/7bf05b57-114b-4c8f-a8fb-2a1ccba8a0e0)
![](https://github.com/godotengine/godot/assets/47700418/6a28e3c5-e512-41c5-ab3c-01ccc02928b8)

To fix the first, you need to replace `use_conversion_assign` with `first_use_conversion_assign` and `second_use_conversion_assign`. See #80247 for reference.

https://github.com/godotengine/godot/blob/be56cab58c056c074d1e02cd0b38641204e39f41/modules/gdscript/gdscript_byte_codegen.cpp#L1618-L1623

To fix the second, you should probably use `OPCODE_ASSIGN` instead of `write_get()`.

4\. Please add more tests.

Also, please remove the reference to the proposal from the commit (leave it only in the first post), because it creates unnecessary noise on each push:

![](https://github.com/godotengine/godot/assets/47700418/2a3f0d70-e85c-4c30-b682-0ba0bc8b9916)

rodrigodias4 commented:
@dalexeev 
1. Changed the grammar and parser to your request, as it makes more sense:
forStmt = "for" IDENTIFIER [ “:” typeHint] [ “,” IDENTIFIER [ “:” typeHint ]] in" expression ":" stmtOrSuite ;

2. Fixed by adding a new OPCODE (OPCODE_FOR_SECOND_VARIABLE_DICT_TYPE_TEST) that checks if the container is a dictionary at runtime before the for loop instead of checking at compile time.

3. Fixed.

4. Added tests akin to the issues you raised.

dalexeev commented on modules/gdscript/gdscript_compiler.cpp:
```diff
@@ -2023,15 +2024,25 @@ Error GDScriptCompiler::_parse_block(CodeGen &codegen, const GDScriptParser::Sui
 
+				if (for_n->second_variable) {
+					second_iterator = codegen.add_local(for_n->second_variable->name, _gdtype_from_datatype(for_n->second_variable->get_datatype(), codegen.script));
+					gen->write_for_dictionary_type_check(second_iterator);
```
Type checking at runtime is not needed if the static analyzer has determined that it is always a dictionary.

<details>
<summary>Details</summary>

```gdscript
func test():
    var a: Dictionary
    for k, v in a: # Runtime type check is not needed.
        pass
    var b: Variant # Runtime type check is needed.
    for k, v in b:
        pass
```

```
Disassembling test()
 0: line 2:     var a: Dictionary
 2: construct validated stack(3) = Dictionary()
 7: line 3:     for k, v in a: # Runtime type check is not needed.
 9: assign stack(7) = stack(3)
 12: type test stack(7) is Dictionary  <-- !!!
 15: for-init (typed DICTIONARY) stack(5) in stack(7) counter stack(6) end 36
 20: jump 27
 22: for-loop (typed DICTIONARY) stack(5) in stack(7) counter stack(6) end 36
 27: get keyed validated stack(9) = stack(7)[stack(6)]
 32: line 4:         pass
 34: jump 22
 36: line 5:     var b: Variant # Runtime type check is needed.
 38: line 6:     for k, v in b:
 40: assign stack(7) = stack(4)
 43: type test stack(7) is Dictionary
 46: for-init stack(5) in stack(7) counter stack(6) end 66
 51: jump 58
 53: for-loop stack(7) in stack(7) counter stack(6) end 66
 58: get keyed stack(9) = stack(7)[stack(6)]
 62: line 7:         pass
 64: jump 53
 66: == END ==
```

</details>

rodrigodias4 replied:
changed to:
```
if (for_n->list->get_datatype().is_variant()) {
    gen->write_for_dictionary_type_check(second_iterator);
}
```
and it seems correct now:

```
func test():
    var a: Dictionary
    for k, v in a: # Runtime type check is not needed.
        pass
    var b: Variant = [1, 2, 3]
    for u, v in b:  # Runtime type check is needed.
        pass
```

> Disassembling test()
>  0: line 2:     var a: Dictionary
>  2: construct validated stack(3) = Dictionary()
>  7: line 3:     for k, v in a: # Runtime type check is not needed.
>  9: assign stack(7) = stack(3)
>  12: for-init (typed DICTIONARY) stack(5) in stack(7) counter stack(6) end 33
>  17: jump 24
>  19: for-loop (typed DICTIONARY) stack(5) in stack(7) counter stack(6) end 33
>  24: get keyed validated stack(9) = stack(7)[stack(6)]
>  29: line 4:         pass
>  31: jump 19
>  33: line 5:     var b: Variant = [1, 2, 3]
>  35:  make_array stack(10) = [const(1), const(2), const(3)]
>  42: assign stack(4) = stack(10)
>  45: assign stack(10) = null
>  47: line 6:     for u, v in b:  # Runtime type check is needed.
>  49: assign stack(7) = stack(4)
>  52: for_loop_second_var_type_test stack(7) is Dictionary
>  54: for-init stack(5) in stack(7) counter stack(6) end 74
>  59: jump 66
>  61: for-loop stack(7) in stack(7) counter stack(6) end 74
>  66: get keyed stack(9) = stack(7)[stack(6)]
>  70: line 7:         pass
>  72: jump 61
>  74: == END ==

dalexeev commented on modules/gdscript/gdscript_vm.cpp:
```diff
@@ -3459,6 +3460,22 @@ Variant GDScriptFunction::call(GDScriptInstance *p_instance, const Variant **p_a
+				CHECK_SPACE(3);
+
+				GET_VARIANT_PTR(container, 0);
+				GET_VARIANT_PTR(second_variable, 1);
```
`second_variable` is unused. Please remove it here, in the codegen (`write_for_dictionary_type_check()`), and in the disassembler. Note that the instruction size will decrease from 3 to 2.

dalexeev commented on modules/gdscript/gdscript_codegen.h:
```diff
@@ -150,6 +150,8 @@ class GDScriptCodeGenerator {
 	virtual void write_end_jump_if_shared() = 0;
 	virtual void start_for(const GDScriptDataType &p_iterator_type, const GDScriptDataType &p_list_type) = 0;
 	virtual void write_for_assignment(const Address &p_list) = 0;
```
```suggestion
	virtual void write_for_first_variable_assignment(const Address &p_list) = 0;
```

And in `gdscript_byte_codegen.h`.

rodrigodias4 replied:
this method is related to the **for** container, not the first variable (and wasn't changed by me). changing to _write_for_container_assignment_ for code clarity

dalexeev commented on modules/gdscript/gdscript_compiler.cpp:
```diff
@@ -2013,6 +2013,7 @@ Error GDScriptCompiler::_parse_block(CodeGen &codegen, const GDScriptParser::Sui
 				codegen.start_block(); // Add an extra block, since the iterator and @special variables belong to the loop scope.
 
 				GDScriptCodeGenerator::Address iterator = codegen.add_local(for_n->variable->name, _gdtype_from_datatype(for_n->variable->get_datatype(), codegen.script));
```
```suggestion
				GDScriptCodeGenerator::Address first_iterator = codegen.add_local(for_n->variable->name, _gdtype_from_datatype(for_n->variable->get_datatype(), codegen.script));
```

dalexeev commented on modules/gdscript/gdscript_vm.cpp:
```diff
@@ -3459,6 +3460,22 @@ Variant GDScriptFunction::call(GDScriptInstance *p_instance, const Variant **p_a
+
+#ifdef DEBUG_ENABLED
+				if (container->get_type() != Variant::DICTIONARY) {
+					err_text = vformat(R"(Unable to assign second variable for iteration of container of type %s (not a dictionary).)", _get_var_type(container));
```
```suggestion
					err_text = vformat(R"(Unable to iterate "%s" as a key-value pair.)", _get_var_type(container));
```

dalexeev commented on modules/gdscript/tests/scripts/parser/features/for_key_value_pair.gd:
```diff
@@ -0,0 +1,3 @@
+func test():
+	for u, v in { "key1" : "value1", "key2" : "value2", "key3" : "value3"}:
```
```suggestion
	for u, v in { "key1": "value1", "key2": "value2", "key3": "value3" }:
```

If you are not testing the parsing, please follow the [code style](https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_styleguide.html). Otherwise, add a comment that you are testing the parser and are intentionally not following the style.

dalexeev commented on modules/gdscript/tests/scripts/parser/features/for_key_value_pair_with_type_specifiers.gd:
```diff
@@ -0,0 +1,3 @@
+func test():
+	for u : String, v : String in { "key1" : "value1", "key2" : "value2", "key3" : "value3"}:
```
```suggestion
	for u: String, v: String in { "key1": "value1", "key2": "value2", "key3": "value3" }:
```

If you are not testing the parsing, please follow the [code style](https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_styleguide.html). Otherwise, add a comment that you are testing the parser and are intentionally not following the style.

dalexeev commented on modules/gdscript/tests/scripts/runtime/errors/for_key_value_incompatible_types.gd:
```diff
@@ -0,0 +1,3 @@
+func test():
+    for u, v: Array in {1: 1, 2: 2}:
```
```suggestion
    for u, v: Array in { 1: 1, 2: 2 }:
```

dalexeev commented on modules/gdscript/tests/scripts/runtime/features/for_key_value_multiple_types.gd:
```diff
@@ -0,0 +1,3 @@
+func test():
+    for u, v in {"key1" : "value1", 2 : "value2", "key3" : 3, 4 : 4, 5.01 : 5.01}:
```
```suggestion
    for u, v in { "key1": "value1", 2: "value2", "key3": 3, 4: 4, 5.01: 5.01 }:
```

dalexeev commented on modules/gdscript/gdscript_disassembler.cpp:
```diff
@@ -996,6 +996,13 @@ void GDScriptFunction::disassemble(const Vector<String> &p_code_lines) const {
 				incr += 3;
 			} break;
+			case OPCODE_FOR_SECOND_VARIABLE_DICT_TYPE_TEST: {
+				text += "type test ";
```
```suggestion
				text += "for_loop_second_var_type_test ";
```

Goldenlion5648 commented:
if the conflicts are resolved, is this able to be merged? The proposal has 89 upvotes

EDIT: I resolved conflicts, see below
